### ISSUE TITLE
Makefile issues

### ISSUE THREAD

**Issue by haampie** [fabiensanglard/st#20]
Due to recent changes:

1. `sudo` is used in install, which makes it impossible to `make install` without sudo on the system, or in build environment where you don't have root privileges, or places where you don't have interactivity to enter a password. Please avoid a build dependency on `sudo`.
2. The relatively standard* `prefix` variable got removed.
3. The `DESTDIR` variable got removed
4. The makefile is no longer portable, it uses `gmake` functions

The use of `DESTDIR` is packaging (see for example [here](https://cmake.org/cmake/help/latest/envvar/DESTDIR.html) and [here](https://www.gnu.org/prep/standards/html_node/DESTDIR.html), so I guess the user in #18 didn't understand the use):

```
make install DESTDIR=./stage/
```

creates `./stage/$prefix` and installs the package there, so you can

```
tar -czf dist.tar.gz -C ./stage .
```

and extract `dist.tar.gz` in `/` on the target filesystem. I think rpms work that way: you package binaries by absolute path, so you need `DESTDIR`. Just `prefix` is insufficient.

\* Please follow the naming scheme from https://www.gnu.org/prep/standards/html_node/Directory-Variables.html instead of doing something custom, that helps. Or more generally, see https://www.gnu.org/prep/standards/html_node/Makefile-Conventions.html.

**Comment by fabiensanglard** [comment]
I was too fast in accepting 6eded70da4f53714cb5d3373db34cc799c36aaba. I did not see that it placed the `.o` next to the source file. It did not like that at all. I asked for a fix it but it was not done so I did it myself as a dev with 1 day experience of `Makefile`.

1. Install is done with `set-user-id`. I don't know how to to that without `sudo`.
2. `prefix`. I had no idea it was standard. I think the current `INSTALLDIR` is clear enough.
3.  `DESTDIR` is gone. Why not `TARGETDIR`?
4. Should it really be portable? The code uses netlink which is Linux only. I was under the impression that all Linux distro had gmake.

**Comment by haampie** [comment]
1. Two things:
   1. Just remove `sudo` from the makefile, and let users run `sudo make install`. That's common enough even without suid, given that `/usr/local` is usually owned by root.
   2. Please separate `install` from `install-setuid` s.t. suid is opt in. (Make them .PHONY and let the latter depend on the former)
2. `INSTALLDIR` is not the same as prefix, since in your case it includes `bin/`. Now users have to look into your makefile to see what it means. And once you add a man page or whatever, you need `prefix` anyways, so why deviate from conventions, there's a reason they exist.


And no, I guess the makefile doesn't have to portable, but then you should maybe document that GNU make is a build dependency.

**Comment by ghost** [comment]
> Install is done with set-user-id. I don't know how to to that without sudo.

Installing a setuid binary doesn't need anything else besides chmod u+s. Based on https://github.com/fabiensanglard/st/commit/863c7f91a937330659e0136ef6b9da32f3312f36:

```sh
[...]
install: all
	mkdir -p $(DESTDIR)$(bindir)
	cp -p ste $(DESTDIR)$(bindir)
	chmod u+s $(DESTDIR)$(bindir)/ste
[...]
```

**Comment by haampie** [comment]
But chown drops the suid bit

**Comment by ghost** [comment]
Maybe we're all overthinking this (myself included). I'm wondering if Linux `capabilities` would offer a way out of setuid for ste.

References:
https://wiki.archlinux.org/title/Security#SUID_and_SGID_files
https://wiki.archlinux.org/title/Capabilities
Capabilities (POSIX 1003.1e, capabilities(7)) provide fine-grained control over superuser permissions, allowing use of the root user to be avoided. Software developers are encouraged to replace uses of the powerful setuid attribute in a system binary with a more minimal set of capabilities. Many packages make use of capabilities, such as CAP_NET_RAW being used for fping. This enables fping to be run by a normal user (as with the setuid method), while at the same time limiting the security consequences of a potential vulnerability in fping.

To summarize: would `CAP_NET_RAW` be enough to let ste do `netlink process monitoring`?

**Comment by haampie** [comment]
So if you make `make install` just copy the file, people can do whatever they want with it post-install.

For setuid / capabilities, you could have a convenience target, like I had in my pr.
